doc/ublk.1
'\" t
.\"     Title: ublk
.\"    Author: [FIXME: author] [see http://www.docbook.org/tdg5/en/html/author]
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 05/03/2025
.\"    Manual: ublk: manage ublk devices
.\"    Source: ublk
.\"  Language: English
.\"
.TH "UBLK" "1" "05/03/2025" "ublk" "ublk: manage ublk devices"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ublk \- Utility to manage ublk devices
.SH "SYNOPSIS"
.HP \w'\fBublk\ <command>\ [<command\ options>]\ [<type\ specific\ options>]\fR\ 'u
\fBublk <command> [<command options>] [<type specific options>]\fR
.SH "DESCRIPTION"
.PP
ublk is a utility that allows you to create, recover, view or delete user\-space block devices\&.
.PP
ublk by default comes with several different types of devices, such as iscsi, nbd, nfs, null and loop\&.
.PP
The following commands are supported:
.SH "ADD COMMAND"
.PP
Command to add a ublk device\&.
.PP
\fB add {\-t, \-\-type} TYPE [{\-n, \-\-number} DEV_ID] [{\-q, \-\-queues} NR_HW_QUEUES] [{\-d, \-\-depth} QUEUE_DEPTH] [{\-u, \-\-uring_comp} URING_COMP] [{\-g, \-\-need\-get\-data} NEED_GET_DATA] [{\-r, \-\-user_recovery} {0|1}] [{\-i, \-\-user_recovery_reissue} {0|1}] [{\-e, \-\-user_recovery_fail_io} {0|1}] [\-\-debug_mask=0x{DBG_MASK}] [\-\-unprivileged] [\-\-cpuset="[SET0],[SET1],\&.\&.\&."] [\-\-usercopy] [{\-z, \-\-zerocopy}] [<type specific options>] \fR
.PP
\fB\-t, \-\-type\fR
.RS 4
Specifies the type of device to create\&. The five types of supported devices are iscsi, nbd, nfs, null and loop\&.
.RE
.PP
\fB\-n, \-\-number\fR
.RS 4
Create a device with this id\&. The device node will be /dev/ublkb_n_\&.
.RE
.PP
\fB\-q, \-\-queue\fR
.RS 4
Number of queues to create\&. Each queue is services by a dedicated child process\&. Default is 1\&.
.RE
.PP
\fB\-d, \-\-depth\fR
.RS 4
Maximum queue\-depthfor each queue\&. Default is 4096\&.
.RE
.PP
\fB\-u, \-\-uring_comp\fR
.RS 4
Force to complete io cmd via io_uring_cmd_complete_in_task so that performance comparison is done easily with using task_work_add\&.
.RE
.PP
\fB\-g, \-\-need_get_data\fR
.RS 4
User should issue io cmd again for write requests to set io buffer address and copy data from bio vectors to the userspace io buffer\&.
.RE
.PP
\fB\-r, \-\-user_recovey\fR
.RS 4
Block devices are recoverable if ublk server exits and restarts\&. Outstanding I/O when ublk server exits is met with errors\&. I/O issued while there is no ublk server queues\&.
.RE
.PP
\fB\-i, \-\-user_recovey_reissue\fR
.RS 4
Block devices are recoverable if ublk server exits and restarts Outstanding I/O when ublk server exits is reissued I/O issued while there is no ublk server queues
.RE
.PP
\fB\-e, \-\-user_recovey_fail_io\fR
.RS 4
Block devices are recoverable if ublk server exits and restarts Outstanding I/O when ublk server exits is met with errors I/O issued while there is no ublk server is met with errors
.RE
.PP
\fB\-\-debug_mask\fR
.RS 4
Bitmask specifying which debug features to enable\&.
.RE
.PP
\fB\-\-unprivileged\fR
.RS 4
Unprivileged user can create /dev/ublkcN and /dev/ublkbN\&.
.sp
/dev/ublk\-control needs to be available for unprivileged user, and it can be done via udev rule to make all control commands available to unprivileged user\&. Except for the command of UBLK_CMD_ADD_DEV, all other commands are only allowed for the owner of the specified device\&.
.sp
When userspace sends UBLK_CMD_ADD_DEV, the device pair\*(Aqs owner_uid and owner_gid are stored to ublksrv_ctrl_dev_info by kernel, so far only the current user\*(Aqs uid/gid is stored, that said owner of the created device is always the current user\&.
.sp
We still need udev rule to apply OWNER/GROUP with the stored owner_uid and owner_gid\&.
.sp
Then ublk server can be run as unprivileged user, and /dev/ublkbN can be accessed and managed by its owner represented by owner_uid/owner_gid\&.
.RE
.PP
\fB\-\-user_copy\fR
.RS 4
Copy between request and user buffer by pread()/pwrite()
.RE
.PP
\fB\-\-cpuset="[SET0],[SET1],\&.\&.\&."\fR
.RS 4
Cpuset is used to specify which cpu(s) a queue will be assigned to\&. By pinning a queue to a specific core(s) that is close to related processing can improve performance\&.
.sp
Each queue has an entry wrapped in \*(Aq[\*(Aq \*(Aq]\*(Aq containing a comma separated list of cores that the cue is assigned to\&.

Example: Create a a loop device with 2 queues where the first queue
is set to use CPUs 3 and 5 and the second que is using CPU 4\&.
  
.sp
.if n \{\
.RS 4
.\}
.nf
    # ublk add \-t loop \-n 0 \-f /tmp/file\&.raw \-\-cpuset="[3,5],[4]"
  
.fi
.if n \{\
.RE
.\}
.RE
.PP
\fB\-z, \-\-zerocopy\fR
.RS 4
Zero\-copy is based on io\-uring uring_cmd of REGISTER_IO_BUF & UNREGISTER_IO_BUF, which avoids data copy between ublk frontend request buffer and ublk server buffer, so memory bandwidth is saved, and throughput & latency improvement can be often observed on large I/O size
.sp
This requires Linux kernel 6\&.15 or later\&.
.RE
.SS "NULL"
.PP
The null device type does not take any extra options\&.
.SS "LOOP"
.PP
Extra options for the loop device type:
.PP
\fB add \-t loop \&.\&.\&. {\-f, \-\-file} FILE [\-\-buffered_io] [\-o, \-\-offset OFFSET] \fR
.PP
\fB\-f, \-\-file\fR
.RS 4
File to use as backing storage for the loop device\&.
.RE
.PP
\fB\-\-buffered_io\fR
.RS 4
Use buffered i/o for accessing the backing file\&. Default is direct i/o\&.
.RE
.PP
\fB\-\-offset\fR
.RS 4
Offset skips first NUM sectors on backing file\&.
.RE
.PP
Example: Create a loop block device
.sp
.if n \{\
.RS 4
.\}
.nf
    # ublk add \-t loop \-n 0 \-f 10M\&.raw
  
.fi
.if n \{\
.RE
.\}
.sp
.SS "NBD"
.PP
Extra options for the nbd (Network Block Device) device type:
.PP
\fB add \-t nbd \&.\&.\&. {\-\-host HOST | \-\-unix UNIX_PATH} \-\-export_name EXP_NAME [\-\-send_zc] [\-\-read_only] \fR
.PP
\fB\-\-host\fR
.RS 4
Hostname of NBD server\&.
.RE
.PP
\fB\-\-host\fR
.RS 4
Hostname of NBD server to use\&.
.RE
.PP
\fB\-\-unix\fR
.RS 4
Path to unix domain socket to use to talk to NBD\&.
.RE
.PP
\fB\-\-export_name\fR
.RS 4
Name of NBD export\&.
.RE
.PP
\fB\-\-send_zc\fR
.RS 4
Use Zero\-Copy\&.
.RE
.PP
\fB\-\-read_only\fR
.RS 4
Read\-only device\&.
.RE
.SS "NFS"
.PP
Extra options for the nfs device type:
.PP
\fB add \-t nfs \&.\&.\&. \-\-nfs NFS\-URL \fR
.PP
\fB\-\-nfs NFS\-URL\fR
.RS 4
URL to the NFS file to use as the block device\&.
.sp
The NFS\-URL format is describe in the libnfs README: https://github\&.com/sahlberg/libnfs/blob/9fa155bfa9d34347a669fbecf4a64259cc573724/README#L55
.RE
.PP
Example: Create a nfs block device
.sp
.if n \{\
.RS 4
.\}
.nf
    # ublk add \-t nfs \-n 0 \-\-nfs nfs://10\&.0\&.0\&.1/export/10M\&.raw
  
.fi
.if n \{\
.RE
.\}
.sp
.SS "iSCSI"
.PP
Extra options for the iSCSI device type\&. iSCSI support requires libiscsi 1\&.20\&.1 or later\&.
.PP
\fB add \-t iscsi \&.\&.\&. \-\-iscsi ISCSI\-URL \-\-initiator\-name NAME \fR
.PP
\fB\-\-iscsi ISCSI\-URL\fR
.RS 4
URL to the iSCSI device to use as the block device\&.
.sp
The ISCSI\-URL format is describe in the libiscsi README: https://github\&.com/sahlberg/libiscsi/blob/eb19863f77e2bad4799ceb90e47fa3bc6205233e/README\&.md?plain=1#L37
.RE
.PP
\fB\-\-initiator\-name NAME\fR
.RS 4
The initiator name to use when logging in to the target\&.
.RE
.PP
Example: Create an iSCSI block device
.sp
.if n \{\
.RS 4
.\}
.nf
    # ublk add \-t iscsi \-n 0 \-\-iscsi iscsi://iscsi\-stgt/iqn\&.2001\-04\&.com\&.ronnie\&.sr0/1 \-\-initiator\-name iqn\&.ronnie\&.test
  
.fi
.if n \{\
.RE
.\}
.sp
.SH "DEL COMMAND"
.PP
Command to delete a ublk device\&.
.PP
\fB del {\-n, \-\-number} DEV_ID [\-a, \-\-all] [\-\-async] \fR
.PP
\fB\-n, \-\-number\fR
.RS 4
Delete the device with this id\&.
.RE
.PP
\fB\-a, \-\-all\fR
.RS 4
Delete all devices\&.
.RE
.PP
Example: Deleting a loop block device
.sp
.if n \{\
.RS 4
.\}
.nf
    # ublk del \-n 0
  
.fi
.if n \{\
.RE
.\}
.sp
.SH "SET_AFFINITY COMMAND"
.PP
Command to change queue affinity\&.
.PP
\fB set_affinity {\-n, \-\-number} DEV_ID [\-q, \-\-queue] QID \-\-cpuset="[SET]" \fR
.PP
\fB\-n, \-\-number\fR
.RS 4
Change the affinity on this device\&.
.RE
.PP
\fB\-q, \-\-queue\fR
.RS 4
Which queue to change the affinity for\&.
.RE
.PP
\fB\-\-cpuset="[SET]"\fR
.RS 4
The new cpuset for this device/queue\&. Format is a comma\-separated list of CPUs within squre brackets\&.
.RE
.PP
Example: Set affinity to core 7 for device 0, queue 1
.sp
.if n \{\
.RS 4
.\}
.nf
    # ublk set_affinity \-n 0 \-q 1 \-\-cpuset="[7]"
  
.fi
.if n \{\
.RE
.\}
.sp
.SH "LIST COMMAND"
.PP
List one or all devices and show their configutaion\&.
.PP
\fB list {\-n, \-\-number} DEV_ID [\-v, \-\-verbose] \fR
.PP
\fB\-n, \-\-number\fR
.RS 4
List the device with this id\&. If omitted all devices will be listed
.RE
.PP
\fB\-v, \-\-verbose\fR
.RS 4
Verbose listing\&. Include the JSON device arguments in the output\&.
.RE
.SH "RECOVER COMMAND"
.PP
Recover a failed ublk device\&.
.PP
\fB recover {\-n, \-\-number} DEV_ID [\-\-cpuset="[SET0],[SET1],\&.\&.\&."] \fR
.PP
\fB\-n, \-\-number\fR
.RS 4
Device to recover\&.
.RE
.PP
\fB\-\-cpuset="[SET0],[SET1],\&.\&.\&."\fR
.RS 4
Cpuset is used to specify which cpu(s) a queue will be assigned to\&. By pinning a queue to a specific core(s) that is close to related processing can improve performance\&.
.sp
Each queue has an entry wrapped in \*(Aq[\*(Aq \*(Aq]\*(Aq containing a comma separated list of cores that the cue is assigned to\&.
.RE
.SH "FEATURES COMMAND"
.PP
Show supported features for the ublk driver\&.
.PP
\fB features \fR
.SH "HELP COMMAND"
.PP
Show generic ot type specific help\&.
.PP
\fB help [{\-t, \-\-type} TYPE] \fR
.PP
\fB\-t, \-\-type\fR
.RS 4
Show help page\&. It \-t is specified, show help page for the specific device type\&.
.RE
.SH "VERSION"
.PP
Show help page\&.\&.
.PP
\fB {\-v, \-\-version} \fR
.SH "RECOVERY"
.PP
There are three arguments that control how ublk will behave in case of a failure, such as crashing\&. The default behavior is no recovery and the device will fail and be removed once the target exists\&.
.PP
To enable recovery mode set "\-\-recovery 1" on the command line\&. Then instead of removing the device upon failure it will instead become inactive in a quiesced state\&.
.sp
.if n \{\
.RS 4
.\}
.nf
dev id 0: nr_hw_queues 1 queue_depth 128 block size 4096 dev_capacity 20480
    max rq size 524288 daemon pid 1239110 state QUIESCED
    flags 0x4a [ URING_CMD_COMP_IN_TASK RECOVERY CMD_IOCTL_ENCODE ]
    ublkc: 511:0 ublkb: 259:4 owner: 0:0
    queue 0: tid 1239112 affinity(0 1 2 3 4 5 6 7 )
    target {"backing_file":"10M","dev_size":10485760,"direct_io":1,"name":"loop","offset":0,"type":0}
  
.fi
.if n \{\
.RE
.\}
.sp
In this state the block device still exists but no I/O can be performed\&.
.PP
To recover a QUIESCED device you can use the recover command:
\fB ublk recover \-n DEV_ID \fR
.PP
There are two additional flags that control how ublk will handle I/O that were in flight when a device is recovered\&.
.SS "\-\-user_recovery_reissue 1"
.PP
When the device is recovered ublk will reissue any I/O that were in flight\&.
.SS "\-\-user_recovery_fail_io 1"
.PP
When the device is recovered ublk will fail all I/O and return an error back to the application\&.
.SH "SEE ALSO"
.PP
\m[blue]\fB\%http://github.com/ublk-org/ublksrv\fR\m[]


ublk.null.cpp
// SPDX-License-Identifier: MIT or GPL-2.0-only

#include <config.h>

#include "ublksrv_tgt.h"

static int null_init_tgt(struct ublksrv_dev *dev, int type, int argc,
		char *argv[])
{
	struct ublksrv_tgt_info *tgt = &dev->tgt;
	const struct ublksrv_ctrl_dev_info *info =
		ublksrv_ctrl_get_dev_info(ublksrv_get_ctrl_dev(dev));
	int jbuf_size;
	char *jbuf = ublksrv_tgt_return_json_buf(dev, &jbuf_size);
	struct ublksrv_tgt_base_json tgt_json = {
		.type = type,
	};
	unsigned long long dev_size = 250UL * 1024 * 1024 * 1024;
	struct ublk_params p = {
		.types = UBLK_PARAM_TYPE_BASIC,
		.basic = {
			.logical_bs_shift	= 9,
			.physical_bs_shift	= 12,
			.io_opt_shift		= 12,
			.io_min_shift		= 9,
			.max_sectors		= info->max_io_buf_bytes >> 9,
			.dev_sectors		= dev_size >> 9,
		},
	};

	strcpy(tgt_json.name, "null");

	if (type != UBLKSRV_TGT_TYPE_NULL)
		return -1;

	if (info->flags & UBLK_F_UNPRIVILEGED_DEV)
		return -1;

	tgt_json.dev_size = tgt->dev_size = dev_size;
	tgt->tgt_ring_depth = info->queue_depth;
	tgt->nr_fds = 0;
	ublksrv_tgt_set_io_data_size(tgt);

	ublk_json_write_dev_info(dev, &jbuf, &jbuf_size);
	ublk_json_write_target_base(dev, &jbuf, &jbuf_size, &tgt_json);
	ublk_json_write_params(dev, &jbuf, &jbuf_size, &p);

	return 0;
}

static int null_recovery_tgt(struct ublksrv_dev *dev, int type)
{
	const struct ublksrv_ctrl_dev *cdev = ublksrv_get_ctrl_dev(dev);
	const char *jbuf = ublksrv_ctrl_get_recovery_jbuf(cdev);
	const struct ublksrv_ctrl_dev_info *info =
		ublksrv_ctrl_get_dev_info(cdev);
	struct ublksrv_tgt_info *tgt = &dev->tgt;
	int ret;
	struct ublk_params p;

	ublk_assert(jbuf);
	ublk_assert(type == UBLKSRV_TGT_TYPE_NULL);

	ret = ublksrv_json_read_params(&p, jbuf);
	if (ret) {
		ublk_err( "%s: read ublk params failed %d\n",
				__func__, ret);
		return ret;
	}

	ublksrv_tgt_set_io_data_size(tgt);
	tgt->dev_size = p.basic.dev_sectors << 9;
	tgt->tgt_ring_depth = info->queue_depth;
	tgt->nr_fds = 0;
	return 0;
}

static co_io_job __null_handle_io_async(const struct ublksrv_queue *q,
		const struct ublk_io_data *data, int tag)
{
	ublksrv_complete_io(q, tag, data->iod->nr_sectors << 9);

	co_return;
}

static int null_handle_io_async(const struct ublksrv_queue *q,
		const struct ublk_io_data *data)
{
	struct ublk_io_tgt *io = __ublk_get_io_tgt_data(data);

	io->co = __null_handle_io_async(q, data, data->tag);

	return 0;
}

struct ublksrv_tgt_type  null_tgt_type = {
	.handle_io_async = null_handle_io_async,
	.init_tgt = null_init_tgt,
	.type	= UBLKSRV_TGT_TYPE_NULL,
	.name	=  "null",
	.recovery_tgt = null_recovery_tgt,
};


static void cmd_usage(const char *name)
{
	printf("ublk.%s -t %s\n", name, name);
	printf("\t-n DEV_ID -q NR_HW_QUEUES -d QUEUE_DEPTH\n");
	printf("\t-u URING_COMP -g NEED_GET_DATA -r USER_RECOVERY\n");
	printf("\t-i USER_RECOVERY_REISSUE -e USER_RECOVERY_FAIL_IO\n");
	printf("\t--debug_mask=0x{DBG_MASK} --unprivileged\n\n");
}

//todo: resolve stack usage warning for mkpath/__mkpath
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wstack-usage="
static int __mkpath(char *dir, mode_t mode)
{
	struct stat sb;
	int ret;
	mode_t mask;

	if (!dir)
		return -EINVAL;

	if (!stat(dir, &sb))
		return 0;

	__mkpath(dirname(strdupa(dir)), mode);

	mask = umask(0);
	ret = mkdir(dir, mode);
	umask(mask);

	return ret;
}

static int mkpath(const char *dir)
{
	return __mkpath(strdupa(dir), S_IRWXU | S_IRWXG | S_IRWXO);
}
#pragma GCC diagnostic pop

static int cmd_dev_add(int argc, char *argv[])
{
	static const struct option longopts[] = {
		{ "type",		1,	NULL, 't' },
		{ "number",		1,	NULL, 'n' },
		{ "queues",		1,	NULL, 'q' },
		{ "depth",		1,	NULL, 'd' },
		{ "zero_copy",		1,	NULL, 'z' },
		{ "uring_comp",		1,	NULL, 'u' },
		{ "need_get_data",	1,	NULL, 'g' },
		{ "user_recovery",	1,	NULL, 'r'},
		{ "user_recovery_fail_io",	1,	NULL, 'e'},
		{ "user_recovery_reissue",	1,	NULL, 'i'},
		{ "debug_mask",	1,	NULL, 0},
		{ "unprivileged",	0,	NULL, 0},
		{ "usercopy",	0,	NULL, 0},
		{ NULL }
	};
	struct ublksrv_dev_data data = {0};
	struct ublksrv_ctrl_dev *dev;
	const struct ublksrv_tgt_type *tgt_type = &null_tgt_type;
	int opt, ret;
	int uring_comp = 0;
	int need_get_data = 0;
	int user_recovery = 0;
	int user_recovery_fail_io = 0;
	int user_recovery_reissue = 0;
	int unprivileged = 0;
	const char *dump_buf;
	int option_index = 0;
	unsigned int debug_mask = 0;

	data.queue_depth = DEF_QD;
	data.nr_hw_queues = DEF_NR_HW_QUEUES;
	data.dev_id = -1;
	data.run_dir = ublksrv_get_pid_dir();

	mkpath(data.run_dir);

	while ((opt = getopt_long(argc, argv, "-:t:n:d:q:u:g:r:e:i:z",
				  longopts, &option_index)) != -1) {
		switch (opt) {
		case 'n':
			data.dev_id = strtol(optarg, NULL, 10);
			break;
		case 't':
			data.tgt_type = optarg;
			break;
		case 'z':
			data.flags |= UBLK_F_SUPPORT_ZERO_COPY;
			break;
		case 'q':
			data.nr_hw_queues = strtol(optarg, NULL, 10);
			break;
		case 'd':
			data.queue_depth = strtol(optarg, NULL, 10);
			break;
		case 'u':
			uring_comp = strtol(optarg, NULL, 10);
			break;
		case 'g':
			need_get_data = strtol(optarg, NULL, 10);
			break;
		case 'r':
			user_recovery = strtol(optarg, NULL, 10);
			break;
		case 'e':
			user_recovery_fail_io = strtol(optarg, NULL, 10);
			break;
		case 'i':
			user_recovery_reissue = strtol(optarg, NULL, 10);
			break;
		case 0:
			if (!strcmp(longopts[option_index].name, "debug_mask"))
				debug_mask = strtol(optarg, NULL, 16);
			if (!strcmp(longopts[option_index].name, "unprivileged"))
				unprivileged = 1;
			if (!strcmp(longopts[option_index].name, "usercopy"))
				data.flags |= UBLK_F_USER_COPY;
			break;
		}
	}

	ublk_set_debug_mask(debug_mask);

	data.max_io_buf_bytes = DEF_BUF_SIZE;
	if (data.nr_hw_queues > MAX_NR_HW_QUEUES)
		data.nr_hw_queues = MAX_NR_HW_QUEUES;
	if (data.queue_depth > MAX_QD)
		data.queue_depth = MAX_QD;
	if (uring_comp)
		data.flags |= UBLK_F_URING_CMD_COMP_IN_TASK;
	if (need_get_data)
		data.flags |= UBLK_F_NEED_GET_DATA;
	if (user_recovery)
		data.flags |= UBLK_F_USER_RECOVERY;
	if (user_recovery_fail_io)
		data.flags |= UBLK_F_USER_RECOVERY | UBLK_F_USER_RECOVERY_FAIL_IO;
	if (user_recovery_reissue)
		data.flags |= UBLK_F_USER_RECOVERY | UBLK_F_USER_RECOVERY_REISSUE;
	if (unprivileged)
		data.flags |= UBLK_F_UNPRIVILEGED_DEV;

	if (data.tgt_type == NULL || strcmp(data.tgt_type, "null")) {
		fprintf(stderr, "Wrong tgt_type specified\n");
		return -EINVAL;
	}

	data.tgt_ops = tgt_type;
	data.flags |= tgt_type->ublk_flags;
	data.ublksrv_flags |= tgt_type->ublksrv_flags;

	//optind = 0;	/* so that tgt code can parse their arguments */
	data.tgt_argc = argc;
	data.tgt_argv = argv;
	dev = ublksrv_ctrl_init(&data);
	if (!dev) {
		fprintf(stderr, "can't init dev %d\n", data.dev_id);
		return -EOPNOTSUPP;
	}

	ret = ublksrv_ctrl_add_dev(dev);
	if (ret < 0) {
		fprintf(stderr, "can't add dev %d, ret %d\n", data.dev_id, ret);
		goto fail;
	}

	{
		const struct ublksrv_ctrl_dev_info *info =
			ublksrv_ctrl_get_dev_info(dev);
		data.dev_id = info->dev_id;
	}
	ret = ublksrv_start_daemon(dev);
	if (ret <= 0) {
		fprintf(stderr, "start dev %d daemon failed, ret %d\n",
				data.dev_id, ret);
		goto fail_del_dev;
	}

	dump_buf = ublksrv_tgt_get_dev_data(dev);
	ublksrv_tgt_set_params(dev, dump_buf);

	ret = ublksrv_ctrl_start_dev(dev, ret);
	if (ret < 0) {
		fprintf(stderr, "start dev %d failed, ret %d\n", data.dev_id,
				ret);
		goto fail_stop_daemon;
	}
	ret = ublksrv_ctrl_get_info(dev);
	ublksrv_ctrl_dump(dev, dump_buf);
	ublksrv_ctrl_deinit(dev);
	return 0;

 fail_stop_daemon:
	ublksrv_stop_io_daemon(dev);
 fail_del_dev:
	ublksrv_ctrl_del_dev(dev);
 fail:
	ublksrv_ctrl_deinit(dev);

	return ret;
}

int main(int argc, char *argv[])
{
	const char *name = "null";
	const char *cmd = "add";
	int ret;

	setvbuf(stdout, NULL, _IOLBF, 0);

	ret = cmd_dev_add(argc, argv);
	if (ret) {
		cmd_usage(name);
		ret = EXIT_FAILURE;
	}

	ublk_ctrl_dbg(UBLK_DBG_CTRL_CMD, "cmd %s: result %d\n", cmd, ret);

	return ret;
}
